### PR TITLE
Refactor reNgine and core task rework

### PR DESCRIPTION
**Pull request by ocervell**
# Overview of the recon process
reNgine reconnaissance workflow

# CHANGELOG

## Task changes

- add a custom `RengineTask` class for engine tasks in `reNgine/celery_custom_task.py`:
The custom task meta-options are toggleable through environment variables:

  `RENGINE_RECORD_ENABLED`:
  - Create / update ScanActivity object to track statuses.
  - Send notifications before and after each task (start / end).
  - Send traceback file to reNgine's Discord channel if an exception happened.
  
  `RENGINE_CACHE_ENABLED`:
  - Get result from cache if it exists.
  - Set result to cache after a task if no exceptions occured.
  
  `RENGINE_RAISE_ON_ERROR`:
  - Raise the actual exception when task fails instead of just logging it.

- Simplify task invocation by passing `ctx` argument instead of `engine_id`, `domain_id`, `scan_history_id`, etc.... The custom `RengineTask` class will convert `ctx` into `self.X` attributes, so that we can use `self.yaml_configuration`, `self.engine_id`, etc... from the tasks themselves.

- Delete a lot of `try` / `except` block in favor of catching exceptions late in the Celery `RengineTask` class. Makes `tasks.py` code way more readable.

- Create a Celery workflow in `reNgine.tasks:initiate_scan` that makes sense (some tasks can be run in parallel, others are better run after others capture some input): this **dramatically decrease the overall run time for a scan** by parallelizing as many tasks as possible. The Celery workflow is now as follows:
reNgine_core

- **Decouple reNgine tasks from each other:**
    - Remove all file-based dependent tasks, use db queries instead
    - Tasks can now be ran completely independently from the others as they query their inputs from the DB instead of files written by previous tasks
    - Do a basic `http_crawl` on the domain URL in the `initiate_scan` function on the base domain URL / IP first so that there is at least 1 endpoint for other tasks to use

- Tasks which fail now send their tracebacks to Discord if `send_status` is enabled. This makes it way easier to debug failed tasks than browsing through hundreds lines of logs.

- Uniformize task names across JS, HTML and Python files: `dir_file_fuzz`, `fetch_url`, `port_scan`, `subdomain_discovery`, `screenshot`, `waf_detection` are the final names.

- Add better support for `SubScan` in all Celery tasks.

- Add a `report` Celery task that takes care of updating `ScanHistory` and generating reports once the scan / subscan is over

- Move `send_hackerone_report` to `tasks.py` and make it a Celery task to allow running it asynchronously

- Add a `remove_duplicate_endpoints` Celery task that checks for similarities in HTTP responses. For now it checks for endpoints which have the same `content_length` or `page_title` (implicit redirects without a `302` response code).

- Add a `stream_command` function that allows to read the output of a command live: this means _**the UI is updated with results while the command runs and does not have to wait until the task completes**_ !!! Re-work most tasks (those that have a JSONLine live output to use `stream_command`  and get results live)

- Use `get_or_create` consistently, remove `query.exists()` calls

- Add functions `save_endpoint`, `save_subdomain`, `save_emails` to add consistency:
    - `save_endpoint` has a `crawl` boolean kwarg that enables to run `http_crawl` (httpx) on it before creation.

- Turn some functions into Celery tasks, so that they can be executed asynchronously to gain time:
    - `http_crawl`
    - `send_notification`
    - `send_discord_message`
    - `send_telegram_message`
    - `send_slack_message`
    - `send_files_to_discord`
    - `query_whois`

- `pwndb` is dead: long live `h8mail` !

- Keep only the **multi-targets tab** in the startScan UI and support any input amongst IPs, URLs, Domains:
    - Add domains (`Domain` saved in DB)
    - Add URLs (`Domain` + `EndPoint` extracted and saved in DB)
    - Add IPs as Domains (`Domain` + `IpAddress` saved in DB)

- Add URL path filtering when initiating a scan. For instance, we want to scan only endpoints starting with `https://example.com/start/`, we can pass the `/start` as a path filter in the startScan UI (new filter tab).

- Better parsing of `gospider` output that follows redirects found in JS etc...

- Add `nmap` support to run vuln scans on ports found by `naabu` (specify `nmap_script` is specified in `port_scan` config)
  - Add `nmap` package to Dockerfile
  - Add `xmltodict` dependency to parse nmap XML results to JSON

- Add shared scan variables (top-level in yaml config):
  - `enable_http_crawl: true`: will enable HTTP crawl. You can disable it to be more stealthy or focus on something different than HTTP.
  - `timeout`: set timeout for all tasks
  - `rate_limit` : set rate limit for all tasks
  - `retries`: set retries for all tasks
  - `custom_header`: set custom header for all tasks

## Notifications

- Add task status notifications for (almost) all tasks: will update Discord embeds with the info found

- Add `send_scan_status_notification` and `send_task_status_notification` functions that simplify updating a task status during the task itself

- Discord notifications: 
  - add / update embed color based on task status, use Redis cache to store previous webhooks / embeds to update
  - add customizable Discord color map tweakable in `web/reNgine/definitions.py`
  - task traceback is now shown in embed when it fails

## Miscellaneous

- Re-work Discord notifications:
    - use embeds for scan status messages: embeds are edited when a scan/subscan/task 's status change
    - add traceback file to embed when the task fails
    - add output file (task results) when the task succeeds

- Change default `NAABU_DEFAULT_PORTS` to `top-1000` instead of `full`, as the latter is **very** time-consuming

## DB  model changes

- Add `EngineType.tasks` hybrid property and remove all boolean `EngineType.dir_file_fuzz`, `EngineType.fetch_url`, etc...
- Change `ScanHistory.celery_id` to `ScanHistory.celery_ids` as a scan can have multiple Celery tasks under it: this allows to stop all sub-Celery tasks when we stop the main scan.
- Change `SubScan.celery_id` to `SubScan.celery_ids`: ditto.

## UI

- Allow showing `SubScan` results while it's running (why not ?)

## Syntax
    - Use f-strings instead of `.format()`
    - Simplify some overly-complicated queries (e.g in `common_func.get_interesting_endpoints` or `common_func.get_interesting_subdomains`)
    - More readable style for queries, for example:
        ```
        subdomain_count = (
            Subdomain.objects
            .filter(...)
            .exclude(...)
            .count()
        )
        ```

## ADMIN

- Add `python3 manage.py loadcustomengines` admin command (load scan engines from files)
- Add `python3 manage.py dumpcustomengines` admin command (dump scan engines to files)

## Related reNgine issues

- Fixes https://github.com/yogeshojha/rengine/issues/200
- Fixes https://github.com/yogeshojha/rengine/issues/249
- Fixes https://github.com/yogeshojha/rengine/issues/500
- Fixes https://github.com/yogeshojha/rengine/issues/568
- Fixes https://github.com/yogeshojha/rengine/issues/574
- Fixes https://github.com/yogeshojha/rengine/issues/582
- Fixes https://github.com/yogeshojha/rengine/issues/625
- Fixes https://github.com/yogeshojha/rengine/issues/629
- Fixes https://github.com/yogeshojha/rengine/issues/658
- Fixes https://github.com/yogeshojha/rengine/issues/678
- Fixes https://github.com/yogeshojha/rengine/issues/687
- Fixes https://github.com/yogeshojha/rengine/issues/692
- Fixes https://github.com/yogeshojha/rengine/issues/693
- Fixes https://github.com/yogeshojha/rengine/issues/698
- Partially fixes https://github.com/yogeshojha/rengine/issues/703

### CODE REVIEW

File: web/api/views.py [hunk 1]
```diff
@@ -1411,7 +1376,7 @@ def paginate_queryset(self, queryset, view=None):
 
 class EndPointChangesViewSet(viewsets.ModelViewSet):
 	'''
-		This viewset will return the EndPoint changes
+		This viewset will return the Endpoint changes
```
Comment (ocervell) [inline web/api/views.py:1379]: TODO: Endpoint -> EndPoint

**Comment by roisec** [comment]
Did you open a pull request for Rengine official repo?

File: default_yaml_config.yaml [hunk 2]
```diff
@@ -27,7 +27,7 @@ dir_file_fuzz:
   threads: 100
   stop_on_error: false
   follow_redirect: false
-  auto_calibration: false
+  auto_calibration: true
```
Comment (ocervell) [inline default_yaml_config.yaml:30]: I changed this to `true` as it really helps remove false positives such as detecting all pages redirected to a 404 page.

**Comment by ocervell** [comment]
@roisec I'm discussing it with @yogeshojha but considering the vast amount of changes we need to review it together and probably bump a major version (2.0.0) for this.

**Pull request merged**